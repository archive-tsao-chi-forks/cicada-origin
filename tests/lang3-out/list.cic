List : (Type) -> Type
List = (T) => { Null(T) | Cons(T) }

Null : Type
Null = {
  kind : "Null" = "Null"
  T : Type
}

Cons : Type
Cons = {
  kind : "Cons" = "Cons"
  T : Type
  head : T
  tail : List(T)
}

null : (T: Type) -> List(T)
null = (T) => {
  kind = "Null"
  T = T
}

cons : (T: Type) -> (T) -> (List(T)) -> List(T)
cons = (T) => (head) => (tail) => {
  kind = "Cons"
  T = T
  head = head
  tail = tail
}

@show null(String)
@show cons(String)("123")(null(String))
