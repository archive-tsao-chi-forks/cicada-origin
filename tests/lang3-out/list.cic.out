{
  kind = "Null"
  T = String
}
{
  kind = "Cons"
  T = String
  head = "123"
  tail = {
    kind = "Null"
    T = String
  }
}

